boot/duneboot.ml

```diff
@@ -189,21 +189,41 @@ let path =
   | exception Not_found -> []
   | s -> split_path s
 
+let find_prog ~f =
+  let rec search = function
+    | [] -> None
+    | dir :: rest -> (
+      match f dir with
+      | None -> search rest
+      | Some fn -> Some (dir, fn))
+  in
+  search path
+
 let exe = if Sys.win32 then ".exe" else ""
 
 (** {2 Concurrency level} *)
 
 let concurrency =
-  let try_run_and_capture_line cmd =
-    let ic, oc, ec = Unix.open_process_full cmd (Unix.environment ()) in
-    let line =
-      match input_line ic with
-      | s -> Some s
-      | exception End_of_file -> None
-    in
-    match (Unix.close_process_full (ic, oc, ec), line) with
-    | WEXITED 0, Some s -> Some s
-    | _ -> None
+  let try_run_and_capture_line (prog, args) =
+    match
+      find_prog ~f:(fun dir ->
+          if Sys.file_exists (dir ^/ prog) then Some prog else None)
+    with
+    | None -> None
+    | Some (dir, prog) -> (
+      let path = dir ^/ prog in
+      let args = Array.of_list @@ (path :: args) in
+      let ic, oc, ec =
+        Unix.open_process_args_full path args (Unix.environment ())
```

Comment:
should this be the empty environment again, if we know the path of the program we're calling?

Reply:
I think clearing the path is generally unintuitive and I'd rather not do it unless there's a good reason. I have no idea how any of these programs I'm running work and if they could be affected by environment variables. It would be safer to just let them pass through.